Add --input-renaming-map and --input-renaming-map-format options

smennetrier opened the pull request:
With these options, it is possible to provide a file
that contain a renaming map in command line.

This feature is useful to build several css files
with the same map (like themes, specific devices, ...).

The generated map contains the content of the input map and
others renaming fields that did not in input map.

smennetrier commented:
Is nobody interested with this feature ? Or everybody is overworked ;)

iflan commented:
Hi Sébastien,

We're interested, but I have some restructuring of our internal repository
to do first. I started last week but ran into some problems that I hope to
address this week. (I thought it would be largely mechanical, but it turned
out that there was quite the hairball lurking under the surface. :-)

Ian
On Mar 19, 2015 3:07 PM, "Sébastien Mennetrier" <notifications@github.com>
wrote:

> Is nobody interested with this feature ? Or everybody is overworked ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/closure-stylesheets/pull/8#issuecomment-83591988>
> .
>

rikuayanokozy commented:
We need this feature. Is everything OK?

smennetrier commented:
I rebased the patch and fix the merge conflict. Everything seems OK for me.

rikuayanokozy commented:
Thanks for your work! It's awesome. I've updated infrastructure of our building system.
And we're still waiting for google accept community patch...

iflan commented:
Hi Sébastien,

I've (finally!) gotten our internal repository ready to accept contributions. If you could update this pull request so that it is on head, that would be great.

Ian

iflan commented:
Hey Sébastien,

If you don't have time for this, I can completely understand. Would you mind if I pulled it in internally, fixed it up and merged it? I'll give you credit in the commit log.

Thanks,

Ian

smennetrier commented:
Hi Ian

Sorry for my silence, I was so busy ;)
I think that I will try to work on this patch this week or next week. I 'll let you know.

If you need urgently, I let you do the merge.

Regards,
Seb

iflan commented:
No worries. This is not urgent—I just didn't want your work to be forgotten.

Thanks!

Ian

smennetrier commented:
Hi Ian,

I rebased my patch on your master branch. I made some modifications according your previous comments.

I didn't create tests case. I don't know how test a method that take a PrintWriter as argument. I have 2 solutions:

- Modify the method writeRenamingMap in RenamingMap Class in order to use a Writer and not a PrintWriter as argument. If the argument is a Writer, we can use the RenamingMap with a StringWriter in tests.

- Create an object that override the PrintWriter object in tests.

I am not a specialist of Java, maybe that others solutions exists. Anyway, I think that the tests can be add in another patch if you are OK.

Let me know if you have others comments on the patch.
Seb

iflan commented:
Hi Sébastien,

Thanks a bunch for doing this! I'm about to take a look at your patch. First, though, you can construct a `PrintWriter` out of any `Writer`, so it should be possible to do something like:
```java
  StringWriter testWriter = new StringWriter();
  PrintWriter testPrintWriter = new PrintWriter(testWriter);
  ...
  CLOSURE_COMPILED.writeRenamingMap(map, testPrintWriter);
  assertEquals("...some expected string...", testWriter.toString());
```
Let me know if that works for you.

Ian

iflan commented on src/com/google/common/css/MultipleMappingSubstitutionMap.java:
```diff
@@ -76,5 +77,5 @@ public static ValueWithMappings createForSingleMapping(String key, String value)
    * {@link ValueWithMappings#mappings}.
    */
-  public ValueWithMappings getValueWithMappings(String key);
+  public ValueWithMappings getValueWithMappings(String key, @Nullable Map<String, String> inputMappings);
```
I'm trying to figure out the intent for adding this parameter. It appears that you're trying to prevent mappings that appear in the input map from being present in the output map. Is that correct?

If so, I think it would it be easier to just post-filter the output map by removing the duplicates. This would even allow us to have this behavior be optional. For example, if you wanted stable renaming, the output map of the previous release would be the input map of the next release.

What do you think?

smennetrier replied:
My first idea was to remove duplicate entries, but it can't work : 

If you have an input map like that : 
{"a": "foo", "b": "bar", ...}

And an output map like that : 
{"a": "foobar", "b": "bar", "c": "foo", ...}

If the entry "a" is removed because it is in two maps. So the "foobar" class name is never renamed.

The result mapping must contain the input map but the result must also use the input mapping if a class names is always renamed and the new generated class names must not be in the input mapping. So I think that you can "post-filter" the output easily.

iflan replied:
I think I'm not understanding something. If you have the input map like:
```javascript
{"foo": "a", "bar": "b", ...}
```
how can you have an output map that maps the keys differently? My understanding is that we would always return keys from the input mapping when they exist.

I guess I'm thinking of a layering like this:
```
SplittingSubstitutionMap -> RecordingSubstitutionMap -> InputMapSubstitutionMap -> ...
```
This means that an input like `foo-quux` would be split first into `foo` and `quux`, then each would go to the `RecordingSubstitutionMap` which would then pass it to the `InputMapSubstitutionMap`. For `foo`, the `InputSubstitutionMap` would return `a` since it is in the input map. The `RecordingSubstitutionMap` would see this and record `"foo": "a"` in its map. For `quux`, `InputMapSubstitutionMap` would not have an entry in the map, so it would call the next `SubstitutionMap`. Assuming this returns a random new ID, `G8ea1` the `RecordingSubstitutionMap` would see that and record `"quux": "G8ea1"` in its map.

I know that `InputMapSubstitutionMap` doesn't exist right now, but that's what I'm proposing to add.

Does this make sense? What am I missing?

Thanks!

smennetrier replied:
I think that this make sense.

And to block the generation of of random ID that exists in input map, Is it ok for you to use an `outputValueBlacklist` in `SubstitutionMap` class (see change `RenamingType.java`) ?

iflan replied:
Oh, I see. I'd forgotten that you might need to know the existing mappings to avoid collisions. `MinimalSubstitutionMap` already has the `outputValueBlacklist`, as you point out, so I think that's a reasonable approach.

smennetrier replied:
Ok, So I will add a class `InputMapSubstitutionMap`.

iflan commented on src/com/google/common/css/RecordingSubstitutionMap.java:
```diff
@@ -68,6 +68,22 @@ public RecordingSubstitutionMap(SubstitutionMap map,
+    this(map, shouldRecordMappingForCodeGeneration);
+
+    if (inputMappings != null) {
+      this.mappings.putAll(inputMappings);
```
I've got to think about this a bit more. You're right that we need to fill the `RecordingSubstitutionMap`'s mappings from the input, but I'm trying to decide whether this is the right place.

smennetrier replied:
Any other ideas ?

smennetrier commented:
Hi Ian,

I think that your idea to construct a ```Printwriter``` with a ```StringWriter``` in parameter should work.

I will try to write test ASAP with this solution :D

smennetrier commented:
Hi Ian,

I pushed a new patch that contain the class `InputMapSubstitutionMap`. I can merge it in previous patch if need.

Let me know if it is OK for you. If it is, I will write tests about that in a new patch.

Seb

iflan commented on src/com/google/common/css/InputMapSubstitutionMap.java:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2009 Google Inc.
```
2015 :-)

iflan commented on src/com/google/common/css/InputMapSubstitutionMap.java:
```diff
@@ -0,0 +1,56 @@
+  private final Map<String,String> mappings;
+
+  /**
+   * Creates a new instance.
```
I know we're inconsistent, but there should be a blank line below this.

iflan commented on src/com/google/common/css/RenamingMapFormat.java:
```diff
@@ -129,20 +150,38 @@ public void writeRenamingMap(Map<String, String> renamingMap,
+  private final String formatPattern;
 
-  private OutputRenamingMapFormat(String formatString) {
+  private RenamingMapFormat(String formatString, String formatPattern) {
```
Since you're re-using this, should we flip the parameter order since the logical order is read, then write?

iflan commented on src/com/google/common/css/RenamingMapFormat.java:
```diff
@@ -80,25 +87,25 @@ public void writeRenamingMap(Map<String, String> renamingMap,
 
   /**
-   * Writes the mapping in a .properties file format, such that it can be read
+   * Read the mapping from a .properties file format, such that it can be read
```
"Read/write" like the others?

iflan commented on src/com/google/common/css/RenamingMapFormat.java:
```diff
@@ -22,41 +22,48 @@
+public enum RenamingMapFormat {
   /**
-   * Writes the mapping as JSON, passed as an argument to
+   * Writes/Read the mapping as JSON, passed as an argument to
```
"Reads/writes"? Here and below.

iflan commented:
Hi Sébastien,

Yes, this looks pretty good. I think that you can write the tests now.

Because this is a pretty big change, I'm going to have to pull this internally and commit it there first. That way I'm sure that I don't break any of our internal clients. So, if you want to just keep adding commits to this branch, that's fine with me. This is the same thing we did with @eunomie's change in #66. You'll still show up as the author for the commit. See 5e681fe82d57942003056786cfb797822ff66502 for an example.

Ian

smennetrier commented:
Hi Ian,

Good :smile:, I will implement tests ASAP and I will push them here.

Seb

smennetrier commented:
Hi Ian,

I'm afraid for my long absence but I was really busy.

I rebased (and so rewrote) my patch in order to add this feature.
I also updated/added some tests.

Let me know if something is missing.

Seb

iflan commented:
Hey Sébastien,

Thanks for this! I'm going to pull this internally to check it out later this week. I'll get back to you after that.

Ian

iflan commented:
Hey Sébastien,

I looked at the change and have pulled it internally for review. I've made some changes to it so that it fits a bit better with how other things are done.

For example, I changed the input format and reader into a `Map` that gets read when the `JobDescription` gets created. That way it's possible to copy the `JobDescription` in a reasonable way.

You'll still get credit, though. :-)

Ian

iflan commented:
Fixed by 13fd65c21baf7c5248570aeb7b22995071de2172.
